// kea-saga/lib/saga/index.js
'use strict';

exports.__esModule = true;

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _effects = require('redux-saga/effects');

var _reduxSaga = require('redux-saga');

var _reduxSaga2 = _interopRequireDefault(_reduxSaga);

var _createSaga = require('./create-saga');

var _createSaga2 = _interopRequireDefault(_createSaga);

var _getConnected = require('./get-connected');

var _getConnected2 = _interopRequireDefault(_getConnected);

var _injectToComponent = require('./inject-to-component');

var _injectToComponent2 = _interopRequireDefault(_injectToComponent);

var _createCombined = require('./create-combined');

var _createCombined2 = _interopRequireDefault(_createCombined);

var _saga = require('./saga');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  name: 'saga',

  // must be used globally
  global: true,
  local: false,

  beforeReduxStore: function beforeReduxStore(options) {
    options._sagaMiddleware = (0, _reduxSaga2.default)();
    options.middleware.push(options._sagaMiddleware);
  },

  afterReduxStore: function afterReduxStore(options, store) {
    options._sagaMiddleware.run(_saga.keaSaga);
    store._sagaMiddleware = options._sagaMiddleware;
  },

  isActive: function isActive(input) {
    return !!(input.sagas || input.start || input.stop || input.takeEvery || input.takeLatest || input.connect && input.connect.sagas);
  },

  afterConnect: function afterConnect(input, output) {
    var connect = input.connect || {};
    var connectedSagas = (0, _getConnected2.default)(connect);

    // sagas we automatically connect from actions && props
    if (connectedSagas.length > 0) {
      output.activePlugins.saga = true;
      input.sagas = input.sagas ? input.sagas.concat(connectedSagas) : connectedSagas;
    }

    // we have input: { connect: { sagas: [] } }, add to input: { sagas: [] }
    if (connect.sagas) {
      input.sagas = input.sagas ? input.sagas.concat(connect.sagas) : connect.sagas;
    }
  },

  afterCreateSingleton: function afterCreateSingleton(input, output) {
    var isActive = output.activePlugins.saga;
    var hasSelectors = !!(output.selectors && (0, _keys2.default)(output.selectors).length > 0);

    if (hasSelectors) {
      output.get = _regenerator2.default.mark(function _callee(key) {
        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return (0, _effects.select)(key ? output.selectors[key] : output.selector);

              case 2:
                return _context.abrupt('return', _context.sent);

              case 3:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      });

      output.fetch = _regenerator2.default.mark(function _callee2() {
        var results,
            keys,
            i,
            _args2 = arguments;
        return _regenerator2.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                results = {};
                keys = Array.isArray(_args2[0]) ? _args2[0] : _args2;
                i = 0;

              case 3:
                if (!(i < keys.length)) {
                  _context2.next = 10;
                  break;
                }

                _context2.next = 6;
                return output.get(keys[i]);

              case 6:
                results[keys[i]] = _context2.sent;

              case 7:
                i++;
                _context2.next = 3;
                break;

              case 10:
                return _context2.abrupt('return', results);

              case 11:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      });
    }

    if (isActive) {
      var singletonSagaBase = {
        constants: (0, _assign2.default)({}, output.constants),
        actions: (0, _assign2.default)({}, output.actions),
        start: input.start,
        stop: input.stop,
        takeEvery: input.takeEvery,
        takeLatest: input.takeLatest,
        workers: input.workers ? (0, _assign2.default)({}, input.workers) : {},
        key: output.key,
        path: output.path,
        get: output.get,
        fetch: output.fetch

        // if saga is a logic store, take it's ".saga", otherwise assume it's a generator function
      };var sagas = (input.sagas || []).map(function (saga) {
        return saga && saga._keaPlugins && saga._keaPlugins.saga && saga.saga || saga;
      });

      if (input.start || input.stop || input.takeEvery || input.takeLatest) {
        output._createdSaga = (0, _createSaga2.default)(singletonSagaBase);
        output.workers = singletonSagaBase.workers;
        sagas.push(output._createdSaga);
      }

      output.saga = _regenerator2.default.mark(function _callee3() {
        var sagaPath;
        return _regenerator2.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                sagaPath = output.path ? output.path.join('.') : input.path('').filter(function (p) {
                  return p;
                }).join('.');
                _context3.next = 3;
                return (0, _effects.call)((0, _createCombined2.default)(sagas, sagaPath));

              case 3:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this);
      });
    }
  },

  injectToClass: function injectToClass(input, output, Klass) {
    if (output.activePlugins.saga) {
      (0, _injectToComponent2.default)(Klass, input, output);
    }
  },

  injectToConnectedClass: function injectToConnectedClass(input, output, KonnektedKlass) {
    if (output.activePlugins.saga) {
      (0, _injectToComponent2.default)(KonnektedKlass, input, output);
    }
  },

  addToResponse: function addToResponse(input, output, response) {
    if (output.activePlugins.saga) {
      response.saga = output.saga;
      response.workers = output.workers;
    }
    response.get = output.get;
    response.fetch = output.fetch;
  }
};
